src/hotspot/share/ci/ciMethod.cpp

```diff
@@ -155,6 +155,10 @@ ciMethod::ciMethod(const methodHandle& h_m, ciInstanceKlass* holder) :
     ciReplay::initialize(this);
   }
 #endif
+
+  if (CompilerOracle::should_blackhole(h_m)) {
+    h_m->set_intrinsic_id(vmIntrinsics::_blackhole);
```

Comment:
Wouldn't it be better to do this in `Method::init_intrinsic_id` so that JVMCI will also see the method as an intrinsic? Either that or a similar bit of logic should be added to `JVMCIRuntime::compile_method`.

Reply:
The way C1 and C2 see this as synthetic intrinsic is modeled after `vmIntrinsics::_compiledLambdaForm`. Unfortunately, `Method::init_intrinsic_id` does not have access to `CompilerOracle` (which is solvable), and current placement in `ciMethod` constructor guarantees the method is resolved. I wonder if JVMCI would benefit from the direct `shouldBlackholeMethod`, like it already has `shouldInlineMethod`?

Reply:
I think it would be better in `JVMCIEnv::get_jvmci_method` which is the JVMCI equivalent of `ciMethod::ciMethod`.

Reply:
It would be best if it was done in ``Method::init_intrinsic_id`` since that's where that field is supposed to be initialized.

Reply:
`Method::init_intrinsic_id` does not have access to `CompilerOracle` and it also does not check for method resolution (I think). I added the similar block to `JVMCIEnv::get_jvmci_method` as @dougxc suggested.